keps/sig-api-machinery/4153-declarative-validation/README.md

```diff
@@ -0,0 +1,1838 @@
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+#### Risk: Migration to Declarative Validation introduce breaking change to API validation
+##### Mitigation: Ensure Invalid Objects Still Invalid
+
+Objects which previously did not pass validation must still not pass validation.
+This can be restated simply as: all currently handwritten validations must be
+included in the schema. If all validations are in the schema, it follows that the
+schema will reject the same set of objects as before.
+
+We can mitigate this by:
+
+1. All Validation Tests Have 100% Coverage
+2. All Validation Tests Succeed against native and schema-based validation backends
+
+To accomplish this, we will:
+
+1. New Presubmit Verify Script to Ensure 100% for future tests
+2. A General Validation Test Framework - Most tests are already written declaratively,
+we will refactor them to all use a general framework for validation tests. The
+framework will evaluate the validation test case first on native types, then convert
+the case to each supported version of the kind to ensure the error arises also
+from schema-based validation. The framework will have a way to make exceptions 
+for certain errors which are version-specific.
+3. Refactor all `validation_tests.go` unit tests to have 100% coverage of their 
+respective `validation.go` files as well as make use of the validation test framework.
+
+##### Mitigation: Ensure Valid Old Objects Still Valid
+
+Objects which previously passed validation must still pass validation. For our
+purposes this can be thought of as ensuring the schema-based validation does not
+raise new errors which previously did not exist.
+
+It is not possible to be 100% sure of this property, in fact, there may be
+new validations added to patch bugs in existing ones. But we can mitigate this 
+risk:
+
+1. Catch and fix any significant differences in schema validations
+  
+  During alpha and beta phases, we will run both validation systems in parallel
+  and return the errors from the old system to the user. We will report in logs
+  and metrics the differences. Over time we will reduce the number of reported
+  differences to zero as these issues are caught.
+
+  To help catch these errors faster, we will also refactor all e2e tests to 
+  enable this feature and check for validation error differences.
+
+2. Minimize the impact of any new changes introduced
+
+  As a precondition of this KEP graduating to GA, the feature 
+  `CRD Validation Ratcheting` must also reach GA. Validation Ratcheting would
+  reduce impact to users by allowing any new schema validations errors due to
+  this feature to be ignored for unchanged portions of an object. After that KEP
+  is GA we will to implement validation ratcheting on native types as part of 
+  this KEP.
```

Comment:
> After that KEP is GA we will to implement validation ratcheting on native types as part of this KEP.

Why can't we implement it while that KEP is in beta?

Reply:
I may have phrased poorly, the main point is "as a precondition of this KEP graduating to GA, the feature 
  `CRD Validation Ratcheting` must also reach GA". The plan is to have ratcheting with this KEP's beta implemention (or possibly alpha, if it can be done cheaply)